consensus/safety-rules/src/safety_rules.rs

```diff
@@ -99,20 +99,30 @@ impl SafetyRules {
     /// 1) B0 <- B1 <- B2 <--
     /// 2) round(B0) + 1 = round(B1), and
     /// 3) round(B1) + 1 = round(B2).
-    pub fn construct_ledger_info(&self, proposed_block: &Block) -> LedgerInfo {
+    pub fn construct_ledger_info(&self, proposed_block: &Block) -> Result<LedgerInfo, Error> {
+        // get rounds of the three blocks
         let block2 = proposed_block.round();
         let block1 = proposed_block.quorum_cert().certified_block().round();
         let block0 = proposed_block.quorum_cert().parent_block().round();
 
-        let commit = block0 + 1 == block1 && block1 + 1 == block2;
-        if commit {
+        // verify 3-chain rule
+        fn next_round(round: u64) -> Result<u64, Error> {
```

Comment:
why not use a lambda `let increment = |round| u64::check_add(round, 1).ok_or(Error::IncorrectRound(round))`

Reply:
I find it less clear personally but I can change it